### PR TITLE
rt haskell refine: make reply functions more consistent

### PR DESCRIPTION
**Pull request by corlewis**
This involves small refactors to replyPop, replyRemove and replyRemoveTCB, and also makes these functions slightly easier to prove properties about. The existing proofs about them are also updated.

### CODE REVIEW

File: proof/refine/ARM/Reply_R.thy [hunk 1]
```diff
@@ -66,33 +69,28 @@ lemma replyUnlink_tcb_obj_at'_no_change:
...
+                            set_reply'.set_no_update[where upd="\<lambda>r. (replyNext_update Map.empty r)"]
+                            set_reply'.get_ko_at')
+         apply (rule_tac Q="\<lambda>_. obj_at' (\<lambda>r. replyTCB r = Some tptr) rptr" in hoare_post_imp,
+                clarsimp)
+         apply (wpsimp wp: hoare_vcg_imp_lift'
+                           set_reply'.set_no_update[where upd="\<lambda>r. (replyPrev_update Map.empty r)"]
+                           set_reply'.get_ko_at')
+        apply (rule_tac Q="\<lambda>_. obj_at' (\<lambda>r. replyTCB r = Some tptr) rptr" in hoare_post_imp,
+               clarsimp)
+        apply (wpsimp wp: gts_wp')+
```
Comment (corlewis) [inline proof/refine/ARM/Reply_R.thy:93]: I'm not happy with this proof but I think a nicer version requires the `updateReply` changes from #95.

**Review by pierzchalski — APPROVED** [review]

**Comment by mktnk3** [comment]
I am not sure if we really want this - the `replyNext` update in `cleanReply` corresponds to setting `reply_sc` to None when the reply is at the head of the reply stack. Swapping `replyUnlink` and `cleanReply` makes the corres proof harder, I think.

**Comment by mktnk3** [comment]
Maybe it is okay with `replyRemove` but probably not for `replyRemoveTCB` and `replyPop`.

**Comment by corlewis** [comment]
> I am not sure if we really want this - the `replyNext` update in `cleanReply` corresponds to setting `reply_sc` to None when the reply is at the head of the reply stack. Swapping `replyUnlink` and `cleanReply` makes the corres proof harder, I think.

Ah, I think you might be right. The main thing that I wanted was for them to all be consistent, it was then just slightly easier for me to prove a few hoare triples if they were in this order. The other order should be fine for me, and it looks like consistently having the replyUnlink at the end of the functions would also be closer to the abstract.

**Comment by mktnk3** [comment]
> Would you be happy with this change if I swapped the order?

Haha, I had to look again to see what would be left if we move `cleanReply` back to where it was ...
So, the only change that remains would be moving down `replyUnlink` to the end of `replyPop`, is this right? Sure, I have no problem with that.

**Comment by corlewis** [comment]
> So, the only change that remains would be moving `replyUnlink` down to the end of `replyPop`, is this right? Sure, I have no problem with that.

Yep, that's right. I'll also update the C, since the order there doesn't always match.

**Comment by corlewis** [comment]
I'll probably end up combining the remaining changes from this with my next PR.

**Pull request closed without merging**